sparta/sparta/functional/RegisterBits.hpp

```diff
@@ -0,0 +1,704 @@
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                Op()(*reinterpret_cast<const SizeT*>(in.remote_data_),
+                     *reinterpret_cast<const SizeT*>(rh_bits.remote_data_));
+            return final_value;
+        }
+
+        template<typename SizeT>
+        static RegisterBits bitNot_(const RegisterBits & in) {
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                std::bit_not<SizeT>()(*reinterpret_cast<const SizeT*>(in.remote_data_));
+            return final_value;
+        }
+
+        template<typename SizeT>
+        static RegisterBits bitShiftRight_(const RegisterBits & in, uint32_t amount) {
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                (*reinterpret_cast<const SizeT*>(in.remote_data_)) >> amount;
+            return final_value;
+        }
+
+        template<typename SizeT>
+        static RegisterBits bitShiftLeft_(const RegisterBits & in, uint32_t amount) {
+            RegisterBits final_value(in.num_bytes_);
+            *reinterpret_cast<SizeT*>(final_value.local_data_) =
+                (*reinterpret_cast<const SizeT*>(in.remote_data_)) << amount;
+            return final_value;
+        }
+
+        // Copy the remote register data locally.
+        void convert_() {
+            if(nullptr == local_data_) {
+                local_data_ = local_storage_.data();
+                ::memset(local_data_, 0, local_storage_.size());
+                ::memcpy(local_data_, remote_data_, num_bytes_);
+                remote_data_ = local_data_;
+            }
+        }
+
+    public:
+        /**
+         * \brief Create an empty class with the given number of bytes
+         * \param num_bytes The number of bytes to allocate
+         */
+        explicit RegisterBits(const uint64_t num_bytes) :
+            local_data_(local_storage_.data()),
+            remote_data_(local_data_),
+            num_bytes_(num_bytes)
+        {
+            ::memset(local_data_, 0, local_storage_.size());
+            sparta_assert(num_bytes <= local_storage_.size(),
+                          "RegisterBits size is locked to 64 bytes. num_bytes requested: " << num_bytes);
+        }
+
+        /**
+         * \brief Create a class with the given number of bytes and initialize it to the given data
+         * \param num_bytes The number of bytes to allocate
+         * \param data The data to write to the lower portion of memory
+         *
+         * The data is copied
+         */
+        template<class DataT>
+        RegisterBits(const uint64_t num_bytes, const DataT & data) :
+            local_storage_(),
+            local_data_(local_storage_.data()),
+            remote_data_(local_data_),
+            num_bytes_(num_bytes)
```

Comment:
Clang bug!  You _have to_ initialize these variables in the member list or clang will barf.  Only happens on template class constructors.  It's been reported.